src/client/wire_protocol/operations.rs

```diff
@@ -0,0 +1,312 @@
+use bson::Document as BSON;
+use bson;
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use client::wire_protocol::header::{Header, OpCode};
+use std::io::{Read, Write};
+use std::mem;
+use std::result::Result::{Ok, Err};
+
+trait ByteLength {
+    /// Calculates the number of bytes in the serialized version of the struct.
+    fn byte_length(&self) -> Result<i32, String>;
+}
+
+impl ByteLength for BSON {
+    /// Gets the length of a BSON document.
+    ///
+    /// # Return value
+    ///
+    /// Returns the number of bytes in the serialized BSON document.
+    fn byte_length(&self) -> Result<i32, String> {
+        let mut temp_buffer = vec![];
+
+        match bson::encode_document(&mut temp_buffer, self) {
+            Ok(_) => Ok(temp_buffer.len() as i32),
+            Err(_) => Err("unable to serialize BSON document".to_owned())
+        }
+    }
+}
+
+/// Represents a message in the MongoDB Wire Protocol.
+pub enum Message {
+    OpReply {
+        header: Header,
+        flags: i32,
+        cursor_id: i64,
+        starting_from: i32,
+        number_returned: i32,
+        documents: Vec<BSON>,
+    },
+    OpQuery {
+        header: Header,
+        flags: i32,
+        full_collection_name: String,
+        number_to_skip: i32,
+        number_to_return: i32,
+        query: BSON,
+        return_field_selector: Option<BSON>,
+    },
+}
+
+impl Message {
+    /// Constructs a new message request for a query.
+    ///
+    /// # Arguments
+    ///
+    /// `header_request_id` - The request ID to be placed in the message header.
+    /// `flags` - Bit vector of query options.
+    /// `full_collection_name` - The full qualified name of the collection,
+    ///                          beginning with the database name and a dot.
+    /// `number_to_skip` - The number of initial documents to skip over in the query
+    ///                    results.
+    /// `number_to_return - The total number of documents that should be returned by
+    ///                     the query.
+    /// `return_field_selector - An optional projection of which fields should be
+    ///                          present in the documents to be returned by the
+    ///                          query.
+    ///
+    /// # Return value
+    ///
+    /// Returns the newly-created Message.
+    pub fn new_query(header_request_id: i32, flags: i32,
```

Comment:
I would also comment with the same as I did in another pull request: this doesn't make a new query, it makes a new message, so I would call it `with_query`